R/sim_contacts_tbl.R

```diff
@@ -0,0 +1,131 @@
+  other_contacts <- subset(
+    other_contacts,
+    subset = !duplicated(other_contacts$infector)
+  )
+
+  # sample contact distribution for non-infected contacts
+  contact_freq <- contact_distribution(nrow(other_contacts))
+
+  # Multiply row by times specified in frequency column V
+  other_contacts <-
+    other_contacts[rep(row.names(other_contacts), contact_freq), ]
+
+  # add random age and gender
+  if (is.data.frame(population_age)) {
+    age_groups <- apply(population_age, MARGIN = 1, function(x) x[1]:x[2])
+    sample_weight <- rep(population_age$proportion, times = lengths(age_groups))
+    # normalise for vector length
+    sample_weight <- sample_weight /
+      rep(lengths(age_groups), times = lengths(age_groups))
+    other_contacts$cnt_age <- sample(
+      unlist(age_groups),
+      size = nrow(other_contacts),
+      replace = TRUE,
+      prob = sample_weight
+    )
+  } else {
+    other_contacts$cnt_age <- sample(
+      population_age[1]:population_age[2],
+      size = nrow(other_contacts),
+      replace = TRUE
+    )
+  }
+  other_contacts$cnt_gender <- sample(
+    c("m", "f"),
+    size = nrow(other_contacts),
+    replace = TRUE
+  )
+
+  # Add corresponding names acc to the gender V
+  other_contacts$to <- randomNames::randomNames(
+    which.names = "both",
+    name.sep = " ",
+    name.order = "first.last",
+    gender = other_contacts$cnt_gender,
+    sample.with.replacement = FALSE
+  )
+
+  # add contact dates
+  other_contacts <- .add_date_last_contact(
+    .data = other_contacts,
+    outbreak_start_date = outbreak_start_date,
+    distribution = config$last_contact_distribution,
+    config$last_contact_distribution_params
+  )
+  other_contacts <- .add_date_first_contact(
+    .data = other_contacts,
+    distribution = config$first_contact_distribution,
+    config$first_contact_distribution_params
+  )
+
+  other_contacts <- subset(
+    other_contacts,
+    select = c(
+      "infector", "from", "to", "cnt_age", "cnt_gender",
+      "date_first_contact", "date_last_contact"
+    )
+  )
+  other_contacts$was_case <- "N"
+
+  # merge both datasets and keep all the values from both
+  contact_investigation <- rbind(contact_investigation, other_contacts)
+  contact_investigation <-
+    contact_investigation[order(contact_investigation$infector), ]
+  row.names(contact_investigation) <- NULL
+
+  # add contact tracing status
+  pick_N <- which(contact_investigation$was_case == "N")
+  status_type <- sample(
+    names(contact_tracing_status_probs),
+    size = length(pick_N),
+    replace = TRUE,
+    prob = contact_tracing_status_probs
+  )
+  contact_investigation$status <- NA
+  contact_investigation$status[pick_N] <- status_type
+  contact_investigation$status <- ifelse(
+    test = is.na(contact_investigation$status),
+    yes = "case",
+    no = contact_investigation$status
+  )
+
+  # remove infector col
+  contact_investigation <- contact_investigation[, -1]
```

Comment:
In general, it's more robust to select/drop by name rather than by position.

Reply:
Resolved in df922daf1416f2a3247062e0bd72a81350c5f671.